index.js

```diff
@@ -124,6 +151,298 @@
+      }
+    };
+    
+    //  addGlue :: a -> Token -> Token -> Token -> Void
+    var addGlue = function(acc, t0, t1, parent) {
+      var glue = rule(parent, 'glue');
+      if (!isEmpty(glue) && !any([t0, t1], isDelimiter)) {
+        insertAfter(parent.tokens, t0, Token('DELIMITER', glue, []));
+      }
+    };
+    
+    //  setStart :: Integer -> Token -> Token -> Integer
+    var setStart = function(acc, token, parent) {
+      token.start = acc;
+      return acc + token.text.length;
+    };
+    
+    //  setEnd :: Integer -> Token -> Token -> Integer
+    var setEnd = function(acc, token, parent) {
+      token.end = isLeaf(token) ? token.start + token.text.length : acc;
+      return token.end;
+    };
+    
+    // create signature from user `def` params
+    return function(name, constraints, expTypes, expRetType) {
+      
+      var sig = Token('SIG', '', [
+        Name(name),
+        Constraints(constraints),
+        Args(expTypes),
+        Ret(expRetType),
+      ]);
+      
+      // delimiters (order matters! bad?)
+      traverse(addPrefix, null, sig);
+      traverse(addWrapper, null, sig);
+      traverse(addSeparator, null, sig);
+      pairTraverse(addGlue, null, sig);
+      
+      // metrics
+      traverse(setStart, 0, sig);
+      postTraverse(setEnd, 0, sig);
+      
+      sig.toString = function() {
+        return traverse(function(acc, t) { return acc.concat(t.text); },
+                        '', sig);
+      };
+      
+      return sig;
+    };
+  })();
+  
+  // for debugging
+  var showToken = function(t) {
+    return !t ? '' : '('+t.type+(t.text ? '='+t.text : '')+')';
+  }
+  
+  //  arrayClone :: [a] -> [a]
+  var arrayClone = function(a) {
+    var b = Array(a.length);
+    for (var idx = 0; idx < a.length; idx += 1) {
+      b[idx] = a[idx];
+    }
+    return b;
+  };
+  
+  var isLeaf = function(token) {
+    return isEmpty(token.tokens);
+  };
+  
+  var isDelimiter = function(t) {
+    return t.type == 'DELIMITER';
+  };
+  
+  var isContainer = function(t) {
+    return t.type == 'UNARY' || t.type == 'BINARY';
+  };
+  
+  var isVar = function(t) {
+    return t.type == 'VARIABLE';
+  };
+  
+  //  traverse :: (((a, b, b) -> a) -> a -> [b] -> a
+  var traverse = function(f, acc, node, parent) {
+    acc = f(acc, node, parent);
+    var tokens = arrayClone(node.tokens);
+    for (var idx = 0; idx < tokens.length; idx += 1) {
+      acc = traverse(f, acc, tokens[idx], node);
+    }
+    return acc;
+  };
+  
+  //  postTraverse :: (((a, b, b) -> a) -> a -> [b] -> a
+  var postTraverse = function(f, acc, node, parent) {
+    var tokens = arrayClone(node.tokens);
+    for (var idx = 0; idx < tokens.length; idx += 1) {
+      acc = postTraverse(f, acc, tokens[idx], node);
+    }
+    return f(acc, node, parent);
+  };
+
+  //  pairTraverse :: (((a, b, b, b) -> a) -> a -> [b] -> a
+  var pairTraverse = function(f, acc, node) {
```

Comment:
This is currently used by `addGlue`. Having this amounts to two 'simple' functions vs using `traverse` for `addGlue` and making that function more complicated.